src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+                return null;
+            }
+        }
+
+        // Core logic to retrieve the local system time zone.
+        // Obtains Android's system local time zone id to get the corresponding time zone
+        // Defaults to Utc if local time zone cannot be found
+        private static TimeZoneInfo GetLocalTimeZoneCore()
+        {
+            string? id = Interop.Sys.GetDefaultTimeZone();
+            if (!string.IsNullOrEmpty(id))
+            {
+                TimeZoneInfo? defaultTimeZone = GetTimeZone(id, id);
+
+                if (defaultTimeZone != null)
+                {
+                    return defaultTimeZone;
+                }
+            }
+
+            return Utc;
+        }
+
+        private static TimeZoneInfoResult TryGetTimeZoneFromLocalMachineCore(string id, out TimeZoneInfo? value, out Exception? e)
+        {
+
+            value = id == LocalId ? GetLocalTimeZoneCore() : GetTimeZone(id, id);
+
+            if (value == null)
+            {
+                e = new InvalidTimeZoneException(SR.Format(SR.InvalidTimeZone_InvalidFileData, id, AndroidTzDataInstance.GetTimeZoneDirectory() + TimeZoneFileName));
+                return TimeZoneInfoResult.TimeZoneNotFoundException;
+            }
+
+            e = null;
+            return TimeZoneInfoResult.Success;
+        }
+
+        private static string[] GetTimeZoneIds()
+        {
+            return AndroidTzDataInstance.GetTimeZoneIds();
+        }
+
+        /*
+        * Android v4.3 Timezone support infrastructure.
+        *
+        * This is a C# port of libcore.util.ZoneInfoDB:
+        *
+        *    https://android.googlesource.com/platform/libcore/+/master/luni/src/main/java/libcore/util/ZoneInfoDB.java
+        *
+        * This is needed in order to read Android v4.3 tzdata files.
+        *
+        * Android 10+ moved the up-to-date tzdata location to a module updatable via the Google Play Store and the
+        * database location changed (https://source.android.com/devices/architecture/modular-system/runtime#time-zone-data-interactions)
+        * The older locations still exist (at least the `/system/usr/share/zoneinfo` one) but they won't be updated.
+        */
+        private sealed class AndroidTzData
+        {
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataHeader
+            {
+                public fixed byte signature[12]; // "tzdata2012f\0"
+                public int indexOffset;
+                public int dataOffset;
+                public int finalOffset;
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
```

Comment:
You should be able to get rid of these `?`s and the subsequent `!` because these arrays are set in the ctor. And if they aren't, then an exception is thrown.

That better models the intent here - these fields are set in the ctor, so you don't need to suppress the nullable warning using `!`s whenever you access them.

Reply:
There is an error when the nullable `?` are removed. 

```
/Users/mdhwang/runtime_droid/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs(226,20): error CS8618: Non-nullable field '_ids' must contain a non-null value when exiting constructor. Consider declaring the field as nullable. [/Users/mdhwang/runtime_droid/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj]
/Users/mdhwang/runtime_droid/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs(226,20): error CS8618: Non-nullable field '_byteOffsets' must contain a non-null value when exiting constructor. Consider declaring the field as nullable. [/Users/mdhwang/runtime_droid/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj]
/Users/mdhwang/runtime_droid/src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs(226,20): error CS8618: Non-nullable field '_lengths' must contain a non-null value when exiting constructor. Consider declaring the field as nullable. [/Users/mdhwang/runtime_droid/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj]
```

I was under the impression that if `LoadData` succeeds, then `LoadTzFile` suceeded, and as such `ReadIndex` succeeded, which sets these fields. 

Could it be that when the exception is thrown that the fields need to be set too?

Reply:
Roslyn doesn't analyze multiple methods at a time. It only analyzes the constructor code to see if `_ids` is set or not.

That's where the `[MemberNotNull]` and `[MemberNotNullWhen]` attributes come in. You can annotate `ReadIndex`, `LoadTzFile`, and `LoadData` with these attributes and the error should go away.

```C#
[MemberNotNull(nameof(_ids)]
private unsafe void ReadIndex(Stream fs, int indexOffset, int dataOffset)

[MemberNotNull(nameof(_ids)]
private unsafe void LoadTzFile(Stream fs)

[MemberNotNullWhen(true, nameof(_ids)]
private bool LoadData(string path)
```

Reply:
Thanks for the insight! I missed the `MemberNotNullWhen` attribute when decorating the methods before reverting back to `?`.